File: LowLevelDesign/java/task-scheduler/src/com/taskscheduler/strategy/RecurringSchedulingStrategy.java
package com.taskscheduler.strategy;

import java.time.Duration;
import java.time.LocalDateTime;
import java.util.Optional;

public class RecurringSchedulingStrategy implements SchedulingStrategy {
    private final Duration interval;
    public RecurringSchedulingStrategy(Duration interval) {
        this.interval = interval;
    }

    @Override
    public Optional<LocalDateTime> getNextExecutionTime(LocalDateTime lastExecutionTime) {
        LocalDateTime baseTime = (lastExecutionTime == null) ? LocalDateTime.now(): lastExecutionTime;
        return Optional.of(baseTime.plus(interval));
    }
}

File: LowLevelDesign/task-scheduler/src/strategy/RecurringSchedulingStrategy.ts
export class RecurringSchedulingStrategy {
    private intervalInMs: number;
    constructor(intervalInMs) {
        this.intervalInMs = intervalInMs;
    }

    getNextExecutionTime(lastExecutionTime){
        let baseTime = new Date()
        if(lastExecutionTime) {
            baseTime = lastExecutionTime
        }
        return new Date(baseTime.getTime() + this.intervalInMs)
    }

}